lib/AST/ASTMangler.cpp

```diff
@@ -1470,48 +1469,61 @@ void ASTMangler::appendFunctionType(AnyFunctionType *fn,
   }
 }
 
-void ASTMangler::appendFunctionSignature(AnyFunctionType *fn,
-                                         bool forceSingleParam) {
-  appendParams(fn->getResult(), /*forceSingleParam*/ false);
-  appendParams(fn->getInput(), forceSingleParam);
+void ASTMangler::appendFunctionSignature(AnyFunctionType *fn) {
+  appendFunctionResultType(fn->getResult());
+  appendFunctionInputType(fn->getParams());
   if (fn->throws())
     appendOperator("K");
 }
 
-void ASTMangler::appendParams(Type ParamsTy, bool forceSingleParam) {
-  if (TupleType *Tuple = ParamsTy->getAs<TupleType>()) {
-    if (Tuple->getNumElements() == 0) {
-      if (forceSingleParam) {
-        // A tuple containing a single empty tuple.
+void ASTMangler::appendFunctionInputType(
+    ArrayRef<AnyFunctionType::Param> params) {
+  switch (params.size()) {
+  case 0:
+    appendOperator("y");
+    break;
+
+  case 1: {
+    auto &param = params[0];
+    if (auto *tupleType = param.getType()->getAs<TupleType>()) {
+      // A paren containing a single empty tuple.
+      if (tupleType->getNumElements() == 0) {
```

Comment:
Why does this need to be a special case? It looks like only the zero-parameter case is special (`y` instead of `yt`).

Reply:
I don't think it does, still trying to figure out everything but i think 1 parameter should always be 

```
appendType(params[0].getType());
if (flags.isShared())
  appendOperator("h");
appendListSeparator();
appendOperator("t");
```

Reply:
well, and 'inout' potentially, and we should leave space for argument labels. That's "just" the patch that's taken for > 1 arguments now.